synapse/storage/databases/main/event_push_actions.py

```diff
@@ -1123,34 +1226,80 @@ def _rotate_notifs_before_txn(
         )
 
         for row in txn:
-            if (row[0], row[1]) in summaries:
-                summaries[(row[0], row[1])].notif_count = row[2]
+            if (row[0], row[1], row[2]) in summaries:
+                summaries[(row[0], row[1], row[2])].notif_count = row[3]
             else:
                 # Because the rules on notifying are different than the rules on marking
                 # a message unread, we might end up with messages that notify but aren't
                 # marked unread, so we might not have a summary for this (user, room)
                 # tuple to complete.
-                summaries[(row[0], row[1])] = _EventPushSummary(
+                summaries[(row[0], row[1], row[2])] = _EventPushSummary(
                     unread_count=0,
-                    stream_ordering=row[3],
-                    notif_count=row[2],
+                    stream_ordering=row[4],
+                    notif_count=row[3],
                 )
 
         logger.info("Rotating notifications, handling %d rows", len(summaries))
 
+        # The upserts need to be partitioned into ones with a thread ID and ones
+        # without.
+        unthreaded_args = [
+            (
+                user_id,
+                room_id,
+                summary.notif_count,
+                summary.unread_count,
+                summary.stream_ordering,
+            )
+            for (user_id, room_id, thread_id), summary in summaries.items()
+            if thread_id is None
+        ]
+
+        threaded_summaries = {
+            (user_id, room_id, thread_id): summary
+            for (user_id, room_id, thread_id), summary in summaries.items()
+            if thread_id is not None
+        }
+
+        # simple_upsert_many_txn doesn't support a predicate clause to force using
+        # the partial index (thread_id = NULL).
```

Comment:
Isn't that easy to add?

Reply:
I tried a bit and was having issues with figuring out how the API should look -- a simple implementation which just accepts a string is pretty easy (I think), but a more complicated one which accepts `keyvalues: Dict[str, Any]` is a bit harder because you need to do conversion of `= NULL` to `IS NULL`. I'm a bit surprised we don't already have code for this?

Anyway, it is doable, yes. I can do it as a separate PR if you'd like.

Reply:
Ah, never mind if it isn't trivial. Was just a bit surprised is all.

Reply:
Looking at it, it should be doable. It would really simplify this code, so will take a look. 👍